Userland/Libraries/LibUSBDB/Database.cpp

```diff
@@ -0,0 +1,228 @@
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/OwnPtr.h>
+#include <AK/RefPtr.h>
+#include <AK/StringView.h>
+
+#include "Database.h"
+
+namespace USBDB {
+
+RefPtr<Database> Database::open(const String& filename)
+{
+    auto file_or_error = MappedFile::map(filename);
+    if (file_or_error.is_error())
+        return nullptr;
+    auto res = adopt_ref(*new Database(file_or_error.release_value()));
+    if (res->init() != 0)
+        return nullptr;
+    return res;
+}
+
+const StringView Database::get_vendor(u16 vendor_id) const
+{
+    const auto& vendor = m_vendors.get(vendor_id);
+    if (!vendor.has_value())
+        return "";
+    return vendor.value()->name;
+}
+
+const StringView Database::get_device(u16 vendor_id, u16 device_id) const
+{
+    const auto& vendor = m_vendors.get(vendor_id);
+    if (!vendor.has_value()) {
+        return "";
+    }
+    const auto& device = vendor.value()->devices.get(device_id);
+    if (!device.has_value())
+        return "";
+    return device.value()->name;
+}
+
+const StringView Database::get_interface(u16 vendor_id, u16 device_id, u16 interface_id) const
+{
+    const auto& vendor = m_vendors.get(vendor_id);
+    if (!vendor.has_value())
+        return "";
+    const auto& device = vendor.value()->devices.get(device_id);
+    if (!device.has_value())
+        return "";
+    const auto& interface = device.value()->interfaces.get(interface_id);
+    if (!interface.has_value())
+        return "";
+    return interface.value()->name;
+}
+
+const StringView Database::get_class(u8 class_id) const
+{
+    const auto& xclass = m_classes.get(class_id);
+    if (!xclass.has_value())
+        return "";
+    return xclass.value()->name;
+}
+
+const StringView Database::get_subclass(u8 class_id, u8 subclass_id) const
+{
+    const auto& xclass = m_classes.get(class_id);
+    if (!xclass.has_value())
+        return "";
+    const auto& subclass = xclass.value()->subclasses.get(subclass_id);
+    if (!subclass.has_value())
+        return "";
+    return subclass.value()->name;
+}
+
+const StringView Database::get_protocol(u8 class_id, u8 subclass_id, u8 protocol_id) const
+{
+    const auto& xclass = m_classes.get(class_id);
+    if (!xclass.has_value())
+        return "";
+    const auto& subclass = xclass.value()->subclasses.get(subclass_id);
+    if (!subclass.has_value())
+        return "";
+    const auto& protocol = subclass.value()->protocols.get(protocol_id);
+    if (!protocol.has_value())
+        return "";
+    return protocol.value()->name;
+}
+
+static u8 parse_hex_digit(char digit)
+{
+    if (digit >= '0' && digit <= '9')
+        return digit - '0';
+    VERIFY(digit >= 'a' && digit <= 'f');
+    return 10 + (digit - 'a');
+}
+
+template<typename T>
+static T parse_hex(StringView str, size_t count)
+{
+    VERIFY(str.length() >= count);
+
+    T res = 0;
+    for (size_t i = 0; i < count; i++)
+        res = (res << 4) + parse_hex_digit(str[i]);
+
+    return res;
+}
```

Comment:
We have half a million of these in the codebase, please use `AK::StringUtils::convert_from_uint_from_hex<T>(StringView)` instead.
`count` is also pointless to take, it's the size of T.

Reply:
Copy and paste from elsewhere in the source has, once again, failed me lol.